/*
Copyright 2017-2019 Echo Park Labs

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

For additional information, contact:

email: info@echoparklabs.io
*/

syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.epl.protobuf";
option java_outer_classname = "GeometryProto";
option objc_class_prefix = "GMPB";
option go_package = "github.com/geo-grpc/api/golang/epl/protobuf";

package epl.protobuf;

message GeometryData {
    int64 geometry_id = 1;
    string feature_id = 2;

    // oneof block
    string wkt = 3;
    string geojson = 4;
    bytes wkb = 5;
    bytes esri_shape = 6;
    // oneof block

    SpatialReferenceData sr = 7;

    EnvelopeData envelope = 8;
}

message SpatialReferenceData {
    // oneof block

    int32 wkid = 1;
    string proj4 = 2;
    string esri_wkt = 3;
    string ogc_wkt = 4;

    // oneof block
}


message EnvelopeData {
    double xmin = 1;
    double ymin = 2;
    double xmax = 3;
    double ymax = 4;
    SpatialReferenceData sr = 5;
}

/*
    type of geometry operators
*/
enum Encoding {
    UNKNOWN_ENCODING = 0; // unknown type.
    WKB = 1; // well-known binary
    WKT = 2; // well-known text
    GEOJSON = 3; // geojson
    ESRI_SHAPE = 4; // esri shape binary
    ENVELOPE = 5;
}

/*
    type of geometry operators
*/
enum OperatorType {
    UNKNOWN_OPERATOR = 0;
    PROJECT = 1;
    EXPORT_TO_JSON = 2;
    IMPORT_FROM_JSON = 3;
    EXPORT_TO_ESRI_SHAPE = 4;
    IMPORT_FROM_ESRI_SHAPE = 5;
    UNION = 6;
    DIFFERENCE = 7;
    PROXIMITY_2D = 8;
    RELATE = 9;
    EQUALS = 10;
    DISJOINT = 11;
    INTERSECTS = 12;
    WITHIN = 13;
    CONTAINS = 14;
    CROSSES = 15;
    TOUCHES = 16;
    OVERLAPS = 17;
    BUFFER = 18;
    DISTANCE = 19;
    INTERSECTION = 20;
    CLIP = 21;
    CUT = 22;
    DENSIFY_BY_LENGTH = 23;
    DENSIFY_BY_ANGLE = 24;
    LABEL_POINT = 25;
    GEODESIC_BUFFER = 26;
    GEODETIC_DENSIFY_BY_LENGTH = 27;

    //    SHAPE_PRESERVING_DENSIFY = 28;

    GEODETIC_LENGTH = 29;
    GEODETIC_AREA = 30;
    SIMPLIFY = 31;
    SIMPLIFY_OGC = 32;
    OFFSET = 33;
    GENERALIZE = 34;
    GENERALIZE_BY_AREA = 35;
    EXPORT_TO_WKB = 36;
    IMPORT_FROM_WKB = 37;
    EXPORT_TO_WKT = 38;
    IMPORT_FROM_WKT = 39;
    IMPORT_FROM_GEOJSON = 40;
    EXPORT_TO_GEOJSON = 41;
    SYMMETRIC_DIFFERENCE = 42;
    CONVEX_HULL = 43;
    BOUNDARY = 44;
    RANDOM_POINTS = 45;
    ENCLOSING_CIRCLE = 46;
}

message GeometryRequest {
    //    oneof primary_geoms {
    GeometryData geometry = 1;
    GeometryData left_geometry = 2;
    GeometryRequest geometry_request = 3;
    GeometryRequest left_geometry_request = 4;
    //    }

    GeometryData right_geometry = 5;
    GeometryRequest right_geometry_request = 6;
    //    }

    OperatorType operator = 7;

    Encoding result_encoding = 8;

    SpatialReferenceData operation_sr = 9;

    SpatialReferenceData result_sr = 10;

    message BufferParams {
        double distance = 1;
        double max_deviation = 2;
        bool union_result = 3;
        int32 max_vertices_in_full_circle = 4;
    }

    message GeodeticBufferParams {
        double distance = 1;
        double max_deviation = 2;
        bool union_result = 3;
        int32 max_vertices_in_full_circle = 4;
    }

    message ConvexParams {
        bool merge = 1;
    }

    message RelateParams {
        string de_9im = 1;
    }

    message RandomPointsParams {
        double points_per_square_km = 1;
        int64 seed = 2;
    }

    message GeneralizeParams {
        double max_deviation = 1;
        bool remove_degenerates = 2;
    }

    message GeneralizeByAreaParams {
        double percent_reduction = 1;
        bool remove_degenerates = 2;
        int32 max_point_count = 3;
    }

    message IntersectionParams {
        int32 dimension_mask = 1;
    }

    message OffsetParams {
        enum OffsetJoinType {
            ROUND = 0;
            BEVEL = 1;
            MITER = 2;
            SQUARE =3;
        }

        double distance = 1;
        OffsetJoinType join_type = 2;
        double bevel_ratio = 3;
        double flatten_error = 4;
    }

    message CutParams {
        bool consider_touch = 1;
    }

    message ClipParams {
        EnvelopeData envelope = 1;
    }

    message DensifyParams {
        double max_length = 1;
    }

    message SimplifyParams {
        bool force = 1;
    }
    //    oneof params {
    BufferParams buffer_params = 11;
    ConvexParams convex_params = 12;
    RelateParams relate_params = 13;
    RandomPointsParams random_points_params = 14;
    GeneralizeParams generalize_params = 15;
    IntersectionParams intersection_params = 16;
    OffsetParams offset_params = 17;
    CutParams cut_params = 18;
    ClipParams clip_params = 19;
    DensifyParams densify_params = 20;
    SimplifyParams simplify_params = 21;
    GeneralizeByAreaParams generalize_by_area_params = 22;
    //    }
}

message GeometryResponse {
    // oneof

    GeometryData geometry = 1;

    bool spatial_relationship = 2;

    double measure = 3;

    map<int64, bool> relate_map = 4;
    EnvelopeData envelope = 5;

    // oneof
}

message FileRequestChunk {
    enum FileType {
        SHAPEFILE = 0;
    }

    GeometryRequest nested_request = 1;
    Encoding result_encoding_type = 2;

    SpatialReferenceData input_sr = 3;
    SpatialReferenceData result_sr = 5;

    bytes data = 6;
    int64 size = 7;
    int64 offset = 8;
    string file_name = 9;

    bool is_last_chunk = 10;
}
